heat/core/io.py

```diff
@@ -354,36 +395,181 @@ def save_netcdf(data, path, variable, mode="w", **kwargs):
+                else:
+                    begin, end, step = begin.flatten(), end.flatten(), step.flatten()
+                    new_slices.append(
+                        np.r_[
+                            tuple(
+                                slice(b.item(), e.item(), s.item())
+                                for b, e, s in zip(begin, end, step)
+                            )
+                        ]
+                    )
+            if out_split is not None:  # add split-slice
+                if isinstance(new_slices[out_split], slice):
+                    start, stop, step = (
+                        new_slices[out_split].start,
+                        new_slices[out_split].stop,
+                        new_slices[out_split].step,
+                    )
+                    sliced = range(start, stop, step)[slices[data.split]]
+                    a, b, c = sliced.start, sliced.stop, sliced.step
+                    a = None if a < 0 else a
+                    b = None if b < 0 else b
+                    new_slices[out_split] = slice(a, b, c)
+                elif isinstance(new_slices[out_split], np.ndarray):
+                    new_slices[out_split] = new_slices[out_split][slices[data.split]]
+                else:
+                    new_slices[out_split] = np.r_[new_slices[out_split]][slices[data.split]]
+            return tuple(new_slices)
+
         # attempt to perform parallel I/O if possible
         if __nc_has_par:
             with nc.Dataset(path, mode, parallel=True, comm=data.comm.handle) as handle:
-                dimension_names = []
-                for dimension, elements in enumerate(data.shape):
-                    name = __NETCDF_DIM_TEMPLATE.format(variable, dimension)
-                    handle.createDimension(name, elements)
-                    dimension_names.append(name)
-
-                var = handle.createVariable(variable, data.dtype.char(), dimension_names, **kwargs)
-                var[slices] = (
-                    data._DNDarray__array.cpu() if is_split else data._DNDarray__array[slices].cpu()
-                )
+                if variable in handle.variables:
+                    var = handle.variables[variable]
+                else:
+                    for name, elements in zip(dimension_names, data.shape):
+                        if name not in handle.dimensions:
+                            handle.createDimension(name, elements if not is_unlimited else None)
+                    var = handle.createVariable(
+                        variable, data.dtype.char(), dimension_names, **kwargs
+                    )
+                mergedSlices = mergeSlices(var, file_slices, data)
+                try:
+                    var[mergedSlices] = (
+                        data._DNDarray__array.cpu()
+                        if is_split
+                        else data._DNDarray__array[slices].cpu()
+                    )
+                except RuntimeError:
```

Comment:
(im unfamiliar with netCDF) why is this raised?

Reply:
Sometimes, collective IO is needed (e.g. for writing into an unlimited dimension). A `RuntimeError` is thrown when attempting this without collective IO.
Someone with more experience in netcdf might know a better convention for when to use collective IO.

Reply:
All metadata operations, e.g. creation, deletion, ..., are collective. This means all processors need to call this operation otherwise the application deadlocks. Should only be necessary in parallel mode to begin with